active-rfcs/0000-component's-public-api.md

```diff
@@ -0,0 +1,408 @@
+
+```html
+<script>
+  export default {
+    expose() {
+      const { foo, bar } = this
+      return {
+        foo, // foo's value is reactive
+        bar
+      }
+    },
+    data() {
+      return {
+        foo: {
+          value: null
+        }
+      }
+    },
+    methods: {
+      bar() {},
+    }
+  }
+</script>
+```
+
+**Composition API example**
+
+```html
+<script>
+  import { ref } from 'vue'
+
+  export default {
+    name: 'MyComponent',
+    setup() {
+      const foo = ref(null)
+      const bar = () => {}
+      return {
+        foo,
+        bar
+      }
+    },
+    expose() {
+      const { foo, bar } = this
+      return {
+        foo, // reactive
+        bar
+      }
+    },
+  }
+</script>
+```
+
+The exposed interface above could be utilized as following:
+
+```html
+<template>
+  <MyComponent :ref="myComponent" />
+</template>
+
+<script>
+  export default {
+    mounted() {
+      console.log(this.$refs.myComponent.foo.value)
+      this.$refs.myComponent.bar()
+    }
+  }
+</script>
+```
+
+To expose refs use `mounted` hook and object as a wrapper to preserve reactivity.
+
+```html
+<template>
+  <input ref="input">
+</template>
+
+<script>
+  export const INPUT_EXPOSED = Symbol()
+  
+  export default {
+    expose() {
+      const { exposed } = this
+      return {
+        [INPUT_EXPOSED]: exposed
+      }
+    },
+    data() {
+      return {
+        exposed: {
+          inputRef: null
+        }
+      }
+    },
+    mounted() {
+      this.exposed.inputRef = this.$refs.input
+    }
+  }
+</script>
+```
+
+Or function refs:
+
+```html
+<template>
+  <input :ref="(input) => exposed.inputRef = input">
+</template>
+
+<script>
+  export const INPUT_EXPOSED = Symbol()
+  
+  export default {
+    expose() {
+      const { exposed } = this
+      return {
+        [INPUT_EXPOSED]: exposed
+      }
+    },
+    data() {
+      return {
+        exposed: {
+          inputRef: null
+        }
+      }
+    },
+  }
+</script>
+```
+
+Alternative way of accessing refs:
+
+```html
+<template>
+  <input ref="input">
+</template>
+
+<script>
+  export const GET_INPUT_REF = Symbol()
+  
+  export default {
+    name: 'MyInput',
+    expose() {
+      return {
+        [GET_INPUT_REF]: () => {
+          return this.$refs.input
+        }
+      }
+    },
+  }
+</script>
+```
+
+```html
+<template>
+  <MyInput ref="input">
+</template>
+
+<script>
+  import { GET_INPUT_REF } from 'MyInput.vue'
+  
+  export default {
+    mounted() {
+      console.log(this.$refs.input[GET_INPUT_REF]())
+    }
+  }
+</script>
+```
+
+# Drawbacks
+
+* `expose` should be executed after data init, `refs` then would require a lot of fiddling around to preserve reactivity (wrapping exposed values with an object at a minimum)
+* Could be difficult to implement
+* Not backwards compatible (but could only warn in compatibility build for example)
```

Comment:
```suggestion
* Not backwards compatible (but could only warn in compatibility build for example)
* Increased bundle size
* Decreased performance: a mechanism like a proxy is necessary to prevent specific properties to be accessible
```

It's worth noting this as well